src/lib/core/Optional.h

```diff
@@ -73,6 +74,15 @@ class Optional
         }
     }
 
+    template <class U, typename = std::enable_if_t<!std::is_same_v<T, U> && std::is_constructible_v<T, const U &>>>
+    constexpr Optional(const Optional<U> & other) : mHasValue(other.HasValue())
```

Comment:
How about making it `explicit` to prevent passing incompatible Optionals mistakenly?

Reply:
The `is_constructible` check makes sure they're compatible, or what do you have in mind in terms of compatibility?

Reply:
I mean that if I have a function that takes `Optional<int>` and I mistakenly pass `Optional<float>`, I would rather have the compiler scream about that :).

Reply:
I've tweaked this now so the Optional constructor is implicit only if the underlying conversion is also implicit, and explicit otherwise. I think this is (similar) to what `std::optional` does (hopefully we can migrate to the std versions soon).